roaring/roaring.go

```diff
@@ -705,93 +674,98 @@ func (b *Bitmap) unionInPlace(others ...*Bitmap) {
-				// want to include the current containers information in the stats.
-				summaryStats = bitmapIters[i:].calculateSummaryStats(iKey)
-			)
-
-			if summaryStats.isOnlyContainerWithKey {
-				// TODO(rartoul): We can avoid these clones if we can determine
-				// if the container is coming from an immutable bitmap and we
-				// know that we can mark the target bitmap as immutable as well.
-				target.Containers.Put(iKey, iContainer.Clone())
-				bitmapIters[i].handled = true
-				continue
-			}
+			iKey, iContainer := iIter.iter.Value()
+			expectedN := int64(0)
 
-			// There was more than one container across the bitmaps with key iKey
-			// so we need to calculate a union.
+			// determine whether we have a target to union into.
+			tContainer := target.Containers.Get(iKey)
+			// if the target's full, short-circuit out.
+			if tContainer != nil {
+				if tContainer.n == maxContainerVal+1 {
+					bitmapIters.markItersWithCurrentKeyAsHandled(i, iKey)
+					continue
+				}
+				expectedN = int64(tContainer.n)
+			}
+			// Check i and later iters for any max-range containers, and
+			// find out how many there are.
+			summaryStats := bitmapIters[i:].calculateSummaryStats(iKey)
 			if summaryStats.hasMaxRange {
 				// One (or more) of the containers represented the maximum possible
 				// range that a container can store, so instead of calculating a
 				// union we can generate an RLE container that represents the entire
 				// range.
-				container := &Container{
+				tContainer = &Container{
 					runs:          []interval16{{start: 0, last: maxContainerVal}},
 					containerType: containerRun,
 					n:             maxContainerVal + 1,
 				}
-				target.Containers.Put(iKey, container)
+				target.Containers.Put(iKey, tContainer)
 				bitmapIters.markItersWithCurrentKeyAsHandled(i, iKey)
 				continue
 			}
-
-			// Use a bitmap container for the target bitmap, and union everything
-			// into it.
-			//
-			// TODO(rartoul): Add another conditional case for n < ArrayMaxSize
-			// (or some fraction of that value) to avoid allocating expensive
-			// bitmaps when unioning many low-density array containers, but this
-			// will require writing a union in place algorithm for an array container
-			// that accepts multiple different containers to union into it for
-			// efficiency.
-			container := target.Containers.Get(iKey)
-			// If target already has a bitmap container for iKey then we can reuse that,
-			// otherwise we have to allocate a new one or convert the existing container
-			// into a bitmap container.
-			if container == nil {
-				buf := make([]uint64, bitmapN)
-				ob := buf[:bitmapN]
-				container = &Container{
-					bitmap:        ob,
-					n:             0,
-					containerType: containerBitmap,
+			expectedN += summaryStats.n
+			var itersToUnion handledIters
+			if tContainer == nil {
+				if summaryStats.c == 1 {
+					// There's no target and we have only one container, we
```

Comment:
There is a lot of if/else nesting in this section which is fine, but it would be a little easier to follow if you began each lexical scope with a little comment like this (specifically you don't have them for a lot of the else statements. I'll leave some examples below.